Plans: Launch unlaunched site when upgrading to atomic plan

jblz opened the pull request:
Note: This is currently targeting the branch in #34193. If that lands first, retarget on master.

As shown in #32671, sites should be launched when upgrading to a plan which takes the site Atomic (at least until we have a good analog for private / unlaunched sites on Atomic).

This change leverages the checkout lib to do just that. It also adds a `Dialog` component to inform the customer that upgrading to plans which will take a site Atomic will make them publicly accessible:

![upgrade-to-atomic-set-to-public](https://user-images.githubusercontent.com/1587282/62673955-524f9880-b96e-11e9-9935-1d17511ea196.gif)

#### Changes proposed in this Pull Request

* Add a `Dialog` signifying that selecting certain plans will immediately launch the site
* Clicking the primary button of the Dialog redirects to the checkout screen
* Set the site to public when a plan which takes the site Atomic is purchased as an upgrade if the site is private
* Add a `QuerySites` component to the Plan features component (so we can reliably check if the site is private / unlaunched)
* Make `isUnlaunchedSite` selector return `false` when no site is selected (to match function documentation saying it should return a bool)


#### Testing instructions

##### Upgrade a Private by default site

* Run this branch
* Browse to `/start`
* Mouse over the "environment badge"
* Mouse over `ABTESTS`
* Scroll down the list of active tests until you see `privateByDefault`
* Click the radio button for `selected`
* Create a Free site (any segment except for online-store)
  * It should be unlaunched ("private by default")
* Go to the plans page
* Click `Business Sites and Online Stores`
* Click the `Upgrade` button for the `Business` or `Online Store` plan
  * You should see a Dialog informing you that your site will be public upon upgrade to this plan
    * Clicking `Cancel` should return you to the plans listing
    * Clicking `Let's Do It` should take you to checkout
      * The appropriate plan should be in your cart
      * You should be able to complete checkout
        * The appropriate upgrades should be in place
* Repeat the above and try to break the flow -- report cases where you can!

##### Upgrade a Private site that's not PbD

* Run this branch
* Browse to `/start`
* Mouse over the "environment badge"
* Mouse over `ABTESTS`
* Scroll down the list of active tests until you see `privateByDefault`
* Click the radio button for `control`
* Create a Free site (any segment except for online-store)
  * It should be public (**NOT** "private by default")
* Change the site to Private
  * `Manage | Settings | Privacy` => `Private`
* Repeat the above -- it should behave identically

##### Upgrade a Public site

* Repeat the previous section, but do not set the site to private -- make sure it's public before the upgrade
* Complete the upgrade to a `Business` or `Online Store` plan
  * You should not see the `Dialog` & the site should remain public

##### Purchase Plans During Signup

* Complete the `/start` signup flow & choose a `Business` or `Online Store` plan
  * You should not see the `Dialog` (for neither email confirmation, nor going public)
  * Your site should be Public after checkout
* Repeat the above -- all other plans should be Private by Default & no `Dialog`s should be shown during signup

##### Test out Jetpack plan purchase & upgrade flows

* Jetpack plan flows should *NOT* see the `Dialog`s & should not otherwise be affected by this change.

Fixes #32671

jblz committed:
Plans: Launch unlaunched site when upgrading to atomic plan

jblz committed:
Move logic to the constructor & skip render

jblz committed:
add site query component back in to ensure the call goes out

p-jackson commented:
Following the testing instructions: when I get to the checkout if I instead click the trash can to cancel the sale then my site will end up launched. I'm still on the free plan, and I can see in the blog report card that it hasn't gone AT. But I was expecting that cancelling would cancel the site launch too.

I might not be a blocker since since the user was warned, but it's not very obvious when the site is launched.

p-jackson commented:
I'll label this **issue 2** for easier communication.

Starting with with a free site I created earlier (before PbD), I go to the settings and switch it to private. Then I look at the plans page.
I was expecting to see the `<Notice>` for the busines and eCommerce plans, but alas they were missing.

The `launch_status` for the site is `false`, which is why `isUnlaunchedSite()` returned `false`.

Maybe there's a difference between "privacy" and "launched-ness" that I'm not aware of though 🤷‍♂

p-jackson commented on client/my-sites/plan-features/index.jsx:
```diff
@@ -851,10 +872,21 @@ export default connect(
 								}
 
+								if ( isUnlaunchedAndGoingAtomic ) {
+									if ( isInSignup ) {
```
Just an observation: I don't think this `isInSignup` check is necessary because everywhere `isInSignup === true`, `onUpgradeClick` has also been provided, so it doesn't go down this branch of the ternary.

jblz replied:
Ya, this is probably unnecessary at the moment, but in case it changes it seems worth leaving the conditional to me as it communicates the intent at the cost of a boolean check.

p-jackson reviewed (commented):
Awesome job putting all this together! There's a lot of stuff to wrangle here!

This PR generally working for me except for the issues I mention above (which might not turn out to be issues).

jblz commented on client/signup/steps/plans/index.jsx:
```diff
@@ -27,14 +27,35 @@ import { getSiteType } from 'state/signup/steps/site-type/selectors';
 import './style.scss';
 
 export class PlansStep extends Component {
+	constructor( props ) {
```
I don't like it any more than you do :)

I lifted the `skipRender` bit from the domains step.  The issue here is with how much of the signup code is contained in [step-actions](https://github.com/Automattic/wp-calypso/blob/a5e45b89533e6ef93a28dc9dde5a27b7576b26e8/client/lib/signup/step-actions.js) which are tightly coupled to the Signup framework. If we could just fire off redux actions to do these sorts of things, a lot of this could be handled at the controller level!

jblz reviewed (commented):
Thanks for taking this for a spin, @p-jackson.  

Ultimately, I'd like for this to not be the way we deal with these upgrades.  Instead, Atomic sites would support being private and unlaunched and this isn't needed.  For now, this is just a workaround that prevents customers who have just upgraded from having to contact support.

monikaburman commented:
Suggestions upon review:
1/ Change the green 'i' to grey; use the UI component pattern as-is.
2/ Copy change: Selecting this plan launches your site.
3/ Link above text to a support article about this feature.

monikaburman reviewed (dismissed):
Suggestions upon review:
1/ Change the green 'i' to grey; use the UI component pattern as-is.
2/ Copy change: Selecting this plan launches your site.
3/ Link above text to a support article about this feature.

jblz commented on client/my-sites/plan-features/index.jsx:
```diff
@@ -537,6 +542,16 @@ export class PlanFeatures extends Component {
 						primaryUpgrade={ primaryUpgrade }
 						selectedPlan={ selectedPlan }
 					/>
+					{ isUnlaunchedAndGoingAtomic && (
```
This needs to apply to private sites across the board -- not just those which are "unlaunched"

jblz replied:
This is done.

jblz committed:
Notice changes:
* Don't set `success` styles -- use default color (gray)
* Copy tweak
* Add an InlineSupportLink

jblz commented:
I pushed some updates to the notice but it looks a bit weird still:

![Screen Shot 2019-07-31 at 1 03 41 PM](https://user-images.githubusercontent.com/1587282/62232203-a4ebec00-b393-11e9-9623-f431d03c60fe.png)

I tried playing around with having the support link just live inline instead of in a `NoticeAction` but it didn't look right that way either.

@michaeldcain is going to see if we can't make it look less weird w/ some styling & / or moving things around a bit.

jblz committed:
Apply to private sites going atomic -- not just unlaunched

jblz committed:
supportPostId as a number for propType validation

monikaburman commented:
> I pushed some updates to the notice but it looks a bit weird still:
> 
> ![Screen Shot 2019-07-31 at 1 03 41 PM](https://user-images.githubusercontent.com/1587282/62232203-a4ebec00-b393-11e9-9623-f431d03c60fe.png)
> 
> I tried playing around with having the support link just live inline instead of in a `NoticeAction` but it didn't look right that way either.
> 


Reducing the text even more: This plan launches your site. 
(Launches your site could be an inline link?)

The 'i' and the '?' are at odds at the two ends of this badge.

jblz committed:
more notice tweaks
* remove notice action and put link as inline text
* not compact

jblz committed:
link on "launches" not "plan"

jblz commented:
OK, this seems nicer:

![launch-for-atomic-upgrade--simpler-notice](https://user-images.githubusercontent.com/1587282/62235659-65290280-b39b-11e9-875c-b2c8b5fda4ea.gif)

Thanks for the suggestions, @monikaburman

a8ci18n commented on client/my-sites/plan-features/index.jsx:
```diff
@@ -540,6 +546,26 @@ export class PlanFeatures extends Component {
+							showDismiss={ false }
+							text={ translate( 'This plan {{launchesLink/}} your site.', {
+								comment:
+									'launchesLink is an inline support link with the (localized) text: "launches"',
```
:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

jblz commented:
Attempting to upgrade without first confirming your email leads to an infinite loop in the launch flow.  We should fix it on both sides, but at least UI block it here.

p-jackson commented on client/my-sites/plan-features/index.jsx:
```diff
@@ -537,6 +542,16 @@ export class PlanFeatures extends Component {
+							icon="info-outline"
+							status="is-success"
+						>
+							{ translate( 'Selecting this plan will launch your site' ) }
```
@jblz sorry I was supposed to submit this yesterday but accidentally left it pending 😬

jblz commented:
I rebased this and updated the test plan..

I'm working on fixing the infinite loop issue when loading the launch flow w/o a confirmed email account now -- or at least UI blocking access when that's the case.

spen commented:
Going straight in to the launch site flow feels cold. If I were a typical Wix-ish user I think I'd prefer just a quick explanation of what was about to happen - Like a very brief splash screen that just says "We're gunna need a few details from you to your launch your new site" with maybe a 'cancel' button and a "Let's go!"

Also, a note on the transitions between the overall signup flow > proper calypso > the launch flow (signup ui) e journey - it feels jarring to me, we're stepping right in and back out of the calypso UI in of another interface environment. I'd love if we still had some reference to it, maybe like a modal, maybe some other modern alternative overlay.

jblz committed:
Plans: Show a Confirmation Dialog When Upgrading Will Launch Site (#35194)

* Plans: Show a Confirmation Dialog When Upgrading Will Launch Site
* UI block upgrade to atomic when email is not verified

jblz commented:
I certainly agree, @spen.  I was really not liking the addition of the `<Notice />` to the features table -- thanks for taking a look here & #35194.

I went ahead and merged the modal / dialog approach into this branch.  That way, there's no way this change will affect the CTR on the `Upgrade` buttons.

I'll add a couple of stat bumps here and I think this is good to merge into the launch PR.

/cc @monikaburman

jblz committed:
Incorporate copy change from #35194

jblz committed:
Send to checkout instead of launch site flow
...& minor styling on the dialog

jblz committed:
Pull out unneeded change to the launch flow / plans step

jblz committed:
pull out unneeded changes to `isUnlaunchedSite`

jblz committed:
signup/plan component from upstream branch

andrewserong commented:
@jblz thanks for detailed testing steps! It's looking really good. I ran into one issue with testing "Upgrade a Private site that's not PbD". I'm running this sandboxed, and if I follow those steps (a private site that isn't private by default) and upgrade to the eCommerce plan, after putting in credit card details in the checkout, I get stuck on the thank you page, and it never updates to the ready state with the CTA to set up your store. An example URL I get stuck on is: `http://calypso.localhost:3000/checkout/thank-you/andysfakeblog201908081526.food.blog/19352358`

In all of the other testing scenarios you've listed, when selecting the eCommerce plan, it takes me to the thank you page and after a time it updates to the "Thank you for your purchase!" state of the thank you page, with the `Create your store!` button.

So unless there's something wrong with my sandbox environment (which is entirely possible, but I reverted to master before testing), it looks like this is somehow to do with a non-private-by-default site that is currently private, and upgrading to the eCommerce plan. (It might not even be directly related to this PR... 🤔)

adamziel commented:
I reproduced the same thing as @andrewserong:

<img width="594" alt="Zrzut ekranu 2019-08-8 o 16 37 32" src="https://user-images.githubusercontent.com/205419/62712296-d6e7f880-b9fa-11e9-94b7-5fc309694a4c.png">

It worked on master branch:

<img width="607" alt="Zrzut ekranu 2019-08-8 o 16 37 55" src="https://user-images.githubusercontent.com/205419/62712320-e2d3ba80-b9fa-11e9-8fec-4738cd80574a.png">

jblz committed:
Defer the site settings call so the thank you page does not get stuck

jblz committed:
Remove Email Verification block -- let them pay us!
Also mv siteIsPrivate to connected props vs plan props.
